src/server/request.rs

```diff
@@ -66,17 +68,31 @@ impl<'a, T> Request<'a, T> {
     #[inline]
     pub fn version(&self) -> &HttpVersion { &self.version }
 
-    /*
     /// The target path of this Request.
     #[inline]
-    pub fn path(&self) -> Option<&str> {
-        match *self.uri {
-            RequestUri::AbsolutePath(ref s) => Some(s),
-            RequestUri::AbsoluteUri(ref url) => Some(&url[::url::Position::BeforePath..]),
-            _ => None
+    pub fn path(&self) -> Option<Cow<str>> {
+        match self.uri {
+            RequestUri::AbsolutePath(ref s) => {
+                let url = Url::parse("http://example.com").ok().and_then(|u| u.join(s).ok());
```

Comment:
Oh ok, good call. I'm wondering whether it's better to run the Url parser
when constructing the RequestUri, and thus be able to return a reference,
or do it on demand, as this patch does.

On Tue, Aug 23, 2016, 7:30 AM Ahmed Charles <notifications@github.com>
wrote:

> In src/server/request.rs
> <https://github.com/hyperium/hyper/pull/899#discussion_r75877722>:
>
> >      /// The target path of this Request.
> >      #[inline]
> > -    pub fn path(&self) -> Option<&str> {
> > -        match *self.uri {
> > -            RequestUri::AbsolutePath(ref s) => Some(s),
> > -            RequestUri::AbsoluteUri(ref url) => Some(&url[::url::Position::BeforePath..]),
> > -            _ => None
> > +    pub fn path(&self) -> Option<Cow<str>> {
> > +        match self.uri {
> > +            RequestUri::AbsolutePath(ref s) => {
> > +                let url = Url::parse("http://example.com").ok().and_then(|u| u.join(s).ok());
>
> I was going for consistency and doing it in a way that I don't have to
> read the RFC and implement correct parsing and interpretation of urls.
> Url::parse normalizes the urls, and therefore, it makes a copy. It would be
> weird if normalized paths and queries were returned for AbsoluteUri but not
> for AbsolutePath.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperium/hyper/pull/899/files/e4ec7e0644b36b81407ec1899036c5c63b1dc2b9#r75877722>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADJF2Xxk2msmuotBxxEluLzk2527Onvks5qiwQMgaJpZM4JqgGd>
> .
>